Services/communication_service.c

```diff
@@ -35,31 +53,753 @@
+                            (uint8_t) packet->data[IN_DATA_BYTE + 3 + 4*i];
+                    U_beacon.MIDI.message[i] = csp_ntoh32(
+                            (uint32_t) U_beacon.MIDI.message[i]);
+                }
+        status = HAL_UHF_setMIDI(U_beacon.MIDI);
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_SET_BEACON_MSG:
+        uhf_struct_len = (uint8_t) packet->data[IN_DATA_BYTE];
+        //uhf_struct_len = 36;
+        U_beacon.message.len = uhf_struct_len;
+        for (i = 0; i < uhf_struct_len; i++)
+        {
+            U_beacon.message.message[i] =
+                    (uint8_t) packet->data[IN_DATA_BYTE + 4 + 4*i];
+            U_beacon.message.message[i] = csp_ntoh32(
+                    (uint32_t) U_beacon.message.message[i]);
+        }
+        status = HAL_UHF_setBeaconMsg(U_beacon.message);
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_SET_I2C:
+        U_I2C_add.add = (uint8_t) packet->data[IN_DATA_BYTE];
+        U_I2C_add.add = csp_ntoh32((uint32_t) U_I2C_add.add);
+        status = HAL_UHF_setI2C(U_I2C_add.add);
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_WRITE_FRAM:
+        cnv8_32(&packet->data[IN_DATA_BYTE], &U_FRAM.add);
+
+        for (i = 0; i < 16; i++)
+        {
+            U_FRAM.data[i] =
+                            (uint8_t) packet->data[IN_DATA_BYTE + 4 + 3 + 4*i];//4 for address
+            U_FRAM.data[i] = csp_ntoh32(
+                            (uint32_t) U_FRAM.data[i]);
+        }
+        status = HAL_UHF_setFRAM(U_FRAM);
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_SECURE:
+        status = HAL_UHF_secure(&U_is_secure);
+        if (sizeof(U_is_secure) + 1 > csp_buffer_data_size())
+        {
+            return SATR_ERROR;
+        }
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &U_is_secure, sizeof(U_is_secure));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(U_is_secure) + 1);
+        if (queue_response(packet) != SATR_OK)
+        {
+            return SATR_ERROR;
+        }
+        break;
+
+    case UHF_GET_FULL_STAT:
```

Comment:
I didn't do a lot of 1-1 for UHF cuz they're too many.

Reply:
I think that's probably fine, I don't imagine those super detailed operations will be used often, since they would require the ground station operator to know a _lot_ about the device